README.md



# geoChina

[![Build Status](https://travis-ci.org/caijun/geoChina.png?branch=master)](https://travis-ci.org/caijun/geoChina)

A R package for geocoding address by calling the Google or Baidu Geocoding API and coordinate conversions for WGS-84, GCJ-02 and BD-09 geodetic system

## Installation

Install the development version from GitHub (`devtools` package is required):

```r
devtools::install_github("geoChina", "caijun")
```


NEWS
Version 1.1.2
-------------------------------------------------------------------------
NEW FEATURES

* minor change to geocode/revgeocode to make function optimized and using friendly .

FIXS

* geocode address out of China with no results returned by calling google map api

DOCUMENTATION CHANGES

* more detail on ics and ocs parameters.



Version 1.1.1
-------------------------------------------------------------------------
FIXS

* revgeocode failed with Error "could not find function "NULLtoNA"" by calling baidu map api
* minor spelling error in DESCRIPTION



Version 1.1
-------------------------------------------------------------------------
NEW FEATURES

* all functions support vectorize operation.

FIXS

* geocode Chinese address failed with status INVALID_REQUEST by calling google map api on Windows platform.
* geocode with turning messaging on failed with Error "invalid multibyte string 1" by calling google map api on Windows platform.

DOCUMENTATION CHANGES

* add examples to illustrate vectorize operation feature.



Version 1.0
-------------------------------------------------------------------------
PACKAGE GENESIS.


R/cst.R
# RGVkaWNhdGVkIHRvIEppYW8gSGFvc29uZy4gSSBsb3ZlIHUh

### WGS-84 => GCJ-02 ###
# Krasovsky 1940 ellipsoid parameters
# semi-major axis
a <- 6378245.0
# inverse flattening: 1/f = 298.3
# flattening
f <- 0.00335233
# semi-minor axis
b <- a * (1 - f)
ee <- (a^2 - b^2) / a^2

#' Convert coordinates
#'
#' converts lat/lon coordintes from WGS-84 to GCJ-02
#' 
#' @param wgsLat a numeric latitude in WGS-84
#' @param wgsLon a numeric longitude in WGS-84
#' @return a data.frame with variables lat/lng
#' @author Jun Cai (\email{cai-j12@@mails.tsinghua.edu.cn}), PhD student from 
#' Center for Earth System Science, Tsinghua University
#' @details for the sake of information security, all real WGS-84 latitude/longitude 
#' coordites must be encrypted by National Admistration of Surveying, Mapping 
#' and Geoinformation (\url{http://en.sbsm.gov.cn/}) into GCJ-02 (known as 'Mars 
#' coordinate system') with a deviation no more than 700 meters in China. Though 
#' the encryption algorithm is highly confidential, the conversion algorithm is 
#' a public secrect on the Internet and verfied to be correct.
#' @seealso \code{\link{gcj2wgs}}, \code{\link{conv}}.
#' 
#' \url{https://on4wp7.codeplex.com/SourceControl/changeset/view/21483#353936} 
#' for C version source code.
#' @export
#' @examples
#' \dontrun{
#' # latitude/longitude coordinates of Beijing railway station
#' # WGS-84: (39.90105, 116.42079)
#' # GCJ-02: (39.90245, 116.42703)
#' wgs2gcj(39.90105, 116.42079) # correct
#' # convert multiple coordinates
#' wgs2gcj(c(39.99837, 39.98565), c(116.3203, 116.2998))
#' }

wgs2gcj <- function(wgsLat, wgsLon){
  # vectorize
  if(length(wgsLat) > 1){
    return(ldply(seq_along(wgsLat), function(i){ wgs2gcj(wgsLat[i], wgsLon[i]) }))
  }
  
  if(outofChina(wgsLat, wgsLon)){
    gcjLat <- wgsLat
    gcjLon <- wgsLat
    return(data.frame(lat = gcjLat, lng = gcjLon))
  }
  
  dLat <- transformLat(wgsLon - 105.0, wgsLat - 35.0)
  dLon <- transformLon(wgsLon - 105.0, wgsLat - 35.0)
  radLat <- wgsLat / 180.0 * pi
  magic <- sin(radLat)
  magic <- 1 - ee * magic * magic
  sqrtMagic <- sqrt(magic)
  dLat <- (dLat * 180.0) / ((a * (1 - ee)) / (magic * sqrtMagic) * pi)
  dLon <- (dLon * 180.0) / (a / sqrtMagic * cos(radLat) * pi)
  gcjLat <- wgsLat + dLat
  gcjLon <- wgsLon + dLon
  return(data.frame(lat = gcjLat, lng = gcjLon))
}

outofChina <- function(lat, lon){
  if(lon < 72.004 | lon > 137.8347) return(TRUE)
  if(lat < 0.8293 | lat > 55.8271) return(TRUE)
  return(FALSE)
}

transformLat <- function(x, y){
  ret <- -100.0 + 2.0 * x + 3.0 * y + 0.2 * y * y + 0.1 * x * y + 0.2 * sqrt(abs(x))
  ret <- ret + (20.0 * sin(6.0 * x * pi) + 20.0 * sin(2.0 * x * pi)) * 2.0 / 3.0
  ret <- ret + (20.0 * sin(y * pi) + 40.0 * sin(y / 3.0 * pi)) * 2.0 / 3.0
  ret <- ret + (160.0 * sin(y / 12.0 * pi) + 320.0 * sin(y * pi / 30.0)) * 2.0 / 3.0
  return(ret)
}

transformLon <- function(x, y){
  ret <- 300.0 + x + 2.0 * y + 0.1 * x * x +  0.1 * x * y + 0.1 * sqrt(abs(x))
  ret <- ret + (20.0 * sin(6.0 * x * pi) + 20.0 * sin(2.0 * x * pi)) * 2.0 / 3.0
  ret <- ret + (20.0 * sin(x * pi) + 40.0 * sin(x / 3.0 * pi)) * 2.0 / 3.0
  ret <- ret + (150.0 * sin(x / 12.0 * pi) + 300.0 * sin(x * pi / 30.0)) * 2.0 / 3.0
  return(ret)
}
### WGS-84 => GCJ-02 ###

### GCJ-02 => WGS-84 ###
#' Convert coordinates
#'
#' converts lat/lon coordintes from GCJ-02 to WGS-84
#' 
#' @param gcjLat a numeric latitude in GCJ-02
#' @param gcjLon a numeric longitude in GCJ-02
#' @return a data.frame with variables lat/lng
#' @author Jun Cai (\email{cai-j12@@mails.tsinghua.edu.cn}), PhD student from 
#' Center for Earth System Science, Tsinghua University
#' @details since the encryption function from WGS-84 to GCJ-02 doesn't have 
#' inverse function, the inverse transformation can only be achieved by numeric 
#' algorithm. Coordinates encryption algorithm must ensure the relative position 
#' correct; that is to say, if point A is close to point B in GCJ-02 system, this 
#' relative spatial relationship persists in WGS-84 system. Therefore, when two 
#' points are close, we can approximate to the right coordinates with the iteration 
#' algorithm.
#' @seealso \code{\link{wgs2gcj}}, \code{\link{conv}}.
#' 
#' the comments of blog \url{http://blog.csdn.net/coolypf/article/details/8686588} 
#' in Chinese
#' @export
#' @examples
#' \dontrun{
#' # latitude/longitude coordinates of Beijing railway station
#' # WGS-84: (39.90105, 116.42079)
#' # GCJ-02: (39.90245, 116.42703)
#' gcj2wgs(39.90245, 116.42703) # correct verifying by google earth
#' # convert multiple coordinates
#' gcj2wgs(c(39.99967, 39.98691), c(116.3264, 116.3059))
#' }

# wgs => gcj
# offset dV = V' - V
# question: gcj => wgs, namely V = V' - dV'
# V' is known, while dV' is unknown.
# since dV is very close to dV', using dV to estimate dV'; however, to calculate
# dV, V must be known. since V and V' are very close to each other, initially 
# using V' to approximate V.
gcj2wgs <- function(gcjLat, gcjLon){
  # vectorize
  if(length(gcjLat) > 1){
    return(ldply(seq_along(gcjLat), function(i){ gcj2wgs(gcjLat[i], gcjLon[i]) }))
  }
  
  g0 <- c(gcjLat, gcjLon)
  w0 <- g0
  g1 <- wgs2gcj(w0[1], w0[2])
  w1 <- w0 - (g1 - g0)
  while(max(abs(w1 - w0)) >= 1e-6){
    w0 <- w1
    g1 <- wgs2gcj(w0[1], w0[2])
    w1 <- w0 - (g1 - g0)
  }
  return(data.frame(lat = w1[1], lng = w1[2]))
}
### GCJ-02 => WGS-84 ###

### GCJ-02 => BD-09 ###
#' Convert coordinates
#'
#' converts lat/lon coordintes from GCJ-02 to BD-09
#' 
#' @param gcjLat a numeric latitude in GCJ-02
#' @param gcjLon a numeric longitude in GCJ-02
#' @return a data.frame with variables lat/lng
#' @author Jun Cai (\email{cai-j12@@mails.tsinghua.edu.cn}), PhD student from 
#' Center for Earth System Science, Tsinghua University
#' @details BD-09 coordinate system is used by Baidu Maps and encrypted based on 
#' GCJ-02 coordinates for information safety.
#' @seealso \code{\link{bd2gcj}}, \code{\link{conv}}.
#' 
#' \url{http://blog.csdn.net/coolypf/article/details/8569813} for C version 
#' source code.
#' @export
#' @examples
#' \dontrun{
#' # latitude/longitude coordinates of Beijing railway station
#' # GCJ-02: (39.90245, 116.42703)
#' # BD-09:  (39.90851, 116.43351)
#' gcj2bd(39.90245, 116.42703) # correct
#' # convert multiple coordinates
#' gcj2bd(c(39.99967, 39.98691), c(116.3264, 116.3059))
#' }

gcj2bd <- function(gcjLat, gcjLon){
  z <- sqrt(gcjLon^2 + gcjLat^2) + 0.00002 * sin(gcjLat * pi * 3000.0 / 180.0)
  theta <- atan2(gcjLat, gcjLon) + 0.000003 * cos(gcjLon * pi * 3000.0 / 180.0)
  bdLon = z * cos(theta) + 0.0065
  bdLat = z * sin(theta) + 0.006
  return(data.frame(lat = bdLat, lng = bdLon))
}
### GCJ-02 => BD-09 ###

### BD-09 => GCJ-02 ###
#' Convert coordinates
#'
#' converts lat/lon coordintes from BD-09 to GCJ-02
#' 
#' @param bdLat a numeric latitude in BD-09
#' @param bdLon a numeric longitude in BD-09
#' @return a data.frame with variables lat/lng
#' @author Jun Cai (\email{cai-j12@@mails.tsinghua.edu.cn}), PhD student from 
#' Center for Earth System Science, Tsinghua University
#' @details BD-09 coordinate system is used by Baidu Maps and encrypted based on 
#' GCJ-02 coordinates for information safety.
#' @seealso \code{\link{gcj2bd}}, \code{\link{conv}}.
#' 
#' \url{http://blog.csdn.net/coolypf/article/details/8569813} for C version 
#' source code.
#' @export
#' @examples
#' \dontrun{
#' # latitude/longitude coordinates of Beijing railway station
#' # GCJ-02: (39.90245, 116.42703)
#' # BD-09:  (39.90851, 116.43351)
#' bd2gcj(39.90851, 116.43351) # correct
#' # convert multiple coordinates
#' bd2gcj(c(40.00541, 39.99299), c(116.3330, 116.3124))
#' }

bd2gcj <- function(bdLat, bdLon){
  x <- bdLon - 0.0065
  y <- bdLat - 0.006  
  z <- sqrt(x^2 + y^2) - 0.00002 * sin(y * pi * 3000.0 / 180.0)
  theta <- atan2(y, x) - 0.000003 * cos(x * pi * 3000.0 / 180.0)  
  gcjLon <- z * cos(theta)  
  gcjLat <- z * sin(theta)
  return(data.frame(lat = gcjLat, lng = gcjLon))
}
### BD-09 => GCJ-02 ###

### BD-09 => GCJ-02 ###
# iteration algorithm same to gcj2wgs function
# bd2gcj <- function(bdLat, bdLon){
#   b0 <- c(bdLat, bdLon)
#   g0 <- b0
#   b1 <- gcj2bd(g0[1], g0[2])
#   g1 <- g0 - (b1 - b0)
#   while(max(abs(g1 - g0)) >= 1e-6){
#     g0 <- g1
#     b1 <- gcj2bd(g0[1], g0[2])
#     g1 <- g0 - (b1 - b0)
#   }
#   return(data.frame(lat = g1[1], lng = g[2]))
# }
### BD-09 => GCJ-02 ###

### WGS-84 <=> BD-09 ###
#' Convert coordinates
#'
#' converts lat/lon coordintes from WGS-84 to BD-09
#' 
#' @param wgsLat a numeric latitude in WGS-84
#' @param wgsLon a numeric longitude in WGS-84
#' @return a data.frame with variables lat/lng
#' @author Jun Cai (\email{cai-j12@@mails.tsinghua.edu.cn}), PhD student from 
#' Center for Earth System Science, Tsinghua University
#' @details convert WGS-84 coordinates first into GCJ-02, then into BD-09.
#' @seealso \code{\link{wgs2gcj}}, \code{\link{gcj2bd}}, \code{\link{conv}}.
#' @export
#' @examples
#' \dontrun{
#' # latitude/longitude coordinates of Beijing railway station
#' # WGS-84: (39.90105, 116.42079)
#' # BD-09:  (39.90851, 116.43351)
#' wgs2bd(39.90105, 116.42079) # correct
#' # convert multiple coordinates
#' wgs2bd(c(39.99837, 39.98565), c(116.3203, 116.2998))
#' }

wgs2bd <- function(wgsLat, wgsLon){
  return(gcj2bd(wgs2gcj(wgsLat, wgsLon)[, 'lat'], wgs2gcj(wgsLat, wgsLon)[, 'lng']))
}

#' Convert coordinates
#'
#' converts lat/lon coordintes from BD-09 to WGS-84
#' 
#' @param bdLat a numeric latitude in BD-09
#' @param bdLon a numeric longitude in BD-09
#' @return a data.frame with variables lat/lng
#' @author Jun Cai (\email{cai-j12@@mails.tsinghua.edu.cn}), PhD student from 
#' Center for Earth System Science, Tsinghua University
#' @details convert BD-09 coordinates first into GCJ-02, then into WGS-84.
#' @seealso \code{\link{bd2gcj}}, \code{\link{gcj2wgs}}, \code{\link{conv}}.
#' @export
#' @examples
#' \dontrun{
#' # latitude/longitude coordinates of Beijing railway station
#' # WGS-84: (39.90105, 116.42079)
#' # BD-09:  (39.90851, 116.43351)
#' bd2wgs(39.90851, 116.43351) # correct verifying by google earth
#' # convert multiple coordinates
#' bd2wgs(c(40.00541, 39.99299), c(116.3330, 116.3124))
#' }

bd2wgs <- function(bdLat, bdLon){
  return(gcj2wgs(bd2gcj(bdLat, bdLon)[, 'lat'], bd2gcj(bdLat, bdLon)[, 'lng']))
}
### WGS-84 <=> BD-09 ###

R/conv.R
#' Convert coordinates
#'
#' the general function that converts lat/lon coordintes from one GCS to another 
#' GCS including WGS-84, GCJ-02 and BD-09 either locally or by calling Baidu 
#' Maps API.
#' 
#' @param lat a numeric latitude
#' @param lon a numeric longitude
#' @param from the inputting GCS
#' @param to the outputting GCS
#' @param api use baidu maps api. Note that baidu maps api only supports the 
#' transformations from WGS-84 or GCJ-02 to BD-09. Other coodinate conversions 
#' must be done locally. As the conversion result is the same, it's recommended 
#' to perform conversions locally.
#' @return a data.frame with variables lat/lng 
#' @author Jun Cai (\email{cai-j12@@mails.tsinghua.edu.cn}), PhD student from 
#' Center for Earth System Science, Tsinghua University
#' @details note that the baidu maps api limits to 20 lat/lon coordinates per query. 
#' Since the coordinate conversion results of Baidu Maps API and local algorithms 
#' are the same, it is recommended to use local algorithms.
#' @seealso \code{\link{wgs2gcj}}, \code{\link{wgs2bd}}, \code{\link{gcj2wgs}}, 
#' \code{\link{gcj2bd}}, \code{\link{bd2wgs}}, \code{\link{bd2gcj}}.
#' @export
#' @import RCurl RJSONIO
#' @examples
#' \dontrun{
#' # latitude/longitude coordinates of Beijing railway station
#' # WGS-84: (39.90105, 116.42079)
#' # GCJ-02: (39.90245, 116.42703)
#' # BD-09:  (39.90851, 116.43351)
#' conv(39.90105, 116.42079, from = 'WGS-84', to = 'GCJ-02')
#' conv(39.90105, 116.42079, from = 'WGS-84', to = 'GCJ-02', api = TRUE)
#' conv(39.90105, 116.42079, from = 'WGS-84', to = 'BD-09')
#' conv(39.90105, 116.42079, from = 'WGS-84', to = 'BD-09', api = TRUE)
#' conv(39.90245, 116.42703, from = 'GCJ-02', to = 'WGS-84')
#' # not supported by baidu map api, return NAs
#' conv(39.90245, 116.42703, from = 'GCJ-02', to = 'WGS-84', api = TRUE)
#' conv(39.90245, 116.42703, from = 'GCJ-02', to = 'BD-09')
#' conv(39.90245, 116.42703, from = 'GCJ-02', to = 'BD-09', api = TRUE)
#' conv(39.90851, 116.43351, from = 'BD-09', to = 'GCJ-02')
#' # not supported by baidu map api, return NAs
#' conv(39.90851, 116.43351, from = 'BD-09', to = 'GCJ-02', api = TRUE)
#' conv(39.90851, 116.43351, from = 'BD-09', to = 'WGS-84')
#' # not supported by baidu map api, return NAs
#' conv(39.90851, 116.43351, from = 'BD-09', to = 'WGS-84', api = TRUE)
#' # convert multiple coordinates
#' lat = c(39.99837, 39.98565)
#' lng = c(116.3203, 116.2998)
#' conv(lat, lng, from = 'WGS-84', to = 'GCJ-02')
#' }

conv <- function(lat, lon, from = c('WGS-84', 'GCJ-02', 'BD-09'), 
                 to = c('WGS-84', 'GCJ-02', 'BD-09'), api = FALSE){  
  # check parameters
  stopifnot(is.numeric(lat))
  stopifnot(is.numeric(lon))
  from <- match.arg(from)
  to <- match.arg(to)
  stopifnot(is.logical(api))
  
  # vectorize
  if(length(lat) > 1){
    return(ldply(seq_along(lat), function(i){conv(lat[i], lon[i], from = from, 
                                                  to = to, api = api) }))
  }
  
  if(from == to){
    return(data.frame(lat = lat, lng = lon))
  } else{
    if(api){
      # coordinate system lookup table
      code <- c(0, 2, 4)
      names(code) <- c('WGS-84', 'GCJ-02', 'BD-09')
      f <- code[from]
      t <- code[to]
      
      # format url
      # http://api.map.baidu.com/ag/coord/convert?x=lon&y=lat&from=FROM&to=TO
      url_string <- paste('http://api.map.baidu.com/ag/coord/convert?x=', lon, 
                          '&y=', lat, '&from=', f, '&to=', t, sep = '')
      url_string <- URLencode(url_string)
      message(paste('calling ', url_string, ' ... ', sep = ''), appendLF = F)
      
      # convert
      connect <- url(url_string)
      cv <- fromJSON(paste(readLines(connect, warn = FALSE), collapse = ''), drop = FALSE)
      message('done.')  
      close(connect)
      
      # did convert fail?
      if(is.list(cv)){
        if(cv$error == 0){
          cvdf <- with(cv, {data.frame(lat = as.numeric(base64(y, FALSE)), 
                                       lng = as.numeric(base64(x, FALSE)), 
                                       row.names = NULL)})
          return(cvdf)
        }
      } else{
        warning(paste('convert failed with error ', cv['error'], 
                      '. conversion is not supported by baidu map api.',
                      sep = ''), 
                call. = FALSE)
        return(data.frame(lat = NA, lng = NA))
      }
    } else{
      if(from == 'WGS-84' & to == 'GCJ-02') return(wgs2gcj(lat, lon))
      if(from == 'WGS-84' & to == 'BD-09') return(wgs2bd(lat, lon))
      if(from == 'GCJ-02' & to == 'WGS-84') return(gcj2wgs(lat, lon))
      if(from == 'GCJ-02' & to == 'BD-09') return(gcj2bd(lat, lon))
      if(from == 'BD-09' & to == 'WGS-84') return(bd2wgs(lat, lon))
      if(from == 'BD-09' & to == 'GCJ-02') return(bd2gcj(lat, lon))
    }
  }
}

R/geocode.R
#' Geocode
#'
#' geocodes an address using Google or Baidu Maps. Note that in most cases by 
#' using this function you are agreeing to the Google Maps API Terms of Service 
#' at \url{https://developers.google.com/maps/terms} or the Baidu Maps API Terms 
#' of Use at \url{http://developer.baidu.com/map/law.htm}.
#' 
#' @param address a character string specifying a location of interest (e.g. 
#' "Tsinghua Univeristy")
#' @param api use google or baidu maps api
#' @param key an api key must be provided when calling baidu maps api. 
#' While it's unnecessary for calling google maps api.
#' @param ocs output coordinate systems including WGS-84, GCJ-02 and BD-09, which
#' are the GCSs of Google Earth, Google Map in China and Baidu Map, respectively. 
#' For address out of China, ocs is automatically set to 'WGS-84' and other values 
#' are igored.
#' @param output lat/lng coordinates or lat/lng coordinates with confidence
#' @param messaging turn messaging on/off
#' @return a data.frame with variables lat/lng or lat/lng/conf 
#' @author Jun Cai (\email{cai-j12@@mails.tsinghua.edu.cn}), PhD student from 
#' Center for Earth System Science, Tsinghua University
#' @details note that the google maps api limits to 2500 queries a day.
#' @seealso \code{\link{revgeocode}}.
#' 
#' Google Maps API at \url{http://code.google.com/apis/maps/documentation/geocoding/} 
#' and Baidu Maps API at \url{http://developer.baidu.com/map/webservice-geocoding.htm}
#' @export
#' @import plyr
#' @examples
#' \dontrun{
#' geocode('Tsinghua University', api = 'google', ocs = 'GCJ-02')
#' geocode('Tsinghua University', api = 'google', ocs = 'WGS-84', 
#'         messaging = TRUE)
#' geocode('Beijing railway station', api = 'google', ocs = 'WGS-84', 
#'         output = 'latlngc')
#' geocode(c('Tsinghua University', 'Beijing railway station'), api = 'google', 
#'         ocs = 'GCJ-02')
#' geocode(c('Tsinghua University', 'Beijing railway station'), api = 'google', 
#'         ocs = 'WGS-84', output = 'latlngc', messaging = TRUE)
#' geocode('Beijing railway station', api = 'baidu', key = 'your baidu maps api key', 
#'         ocs = 'BD-09')
#' geocode('Beijing railway station', api = 'baidu', key = 'your baidu maps api key', 
#'         ocs = 'GCJ-02', messaging = TRUE)
#' geocode('Beijing railway station', api = 'baidu', key = 'your baidu maps api key', 
#'         ocs = 'BD-09', output = 'latlngc')
#' geocode(c('Tsinghua University', 'Beijing railway station'), api = 'baidu', 
#'         key = 'your baidu maps api key', ocs = 'BD-09')
#' geocode(c('Tsinghua University', 'Beijing railway station'), api = 'baidu', 
#'         key = 'your baidu maps api key', ocs = 'WGS-84', output = 'latlngc')
#' }

geocode <- function(address, api = c('google', 'baidu'), key = '', 
                    ocs = c('WGS-84', 'GCJ-02', 'BD-09'), 
                    output = c('latlng', 'latlngc'), messaging = FALSE){  
  # check parameters
  stopifnot(is.character(address))
  api <- match.arg(api)
  stopifnot(is.character(key))
  output <- match.arg(output)
  ocs <- match.arg(ocs)
  stopifnot(is.logical(messaging))
  
  # vectorize for many addresses
  if(length(address) > 1){
    if(api == 'google'){
      s <- 'google restricts requests to 2500 requests a day.'
      if(length(address) > 2500) stop(s, call. = F)
      if(length(address) > 200 & messaging) message(paste('Reminder', s, sep = ' : '))
    }
          
    if(output == 'latlng' | output == 'latlngc'){
      return(ldply(address, geocode, api = api, key = key, ocs = ocs, 
                   output = output, messaging = messaging))
    }
  }
  
  # location encoding
  address <- enc2utf8(address)
  
  # format url
  if(api == 'google'){
    # https is only supported on Windows, when R is started with the --internet2 
    # command line option. without this option, or on Mac, you will get the error 
    # "unsupported URL scheme".
    # http://maps.googleapis.com/maps/api/geocode/json?address=ADDRESS&sensor
    # =false&key=API_KEY
    url_string <- paste('http://maps.googleapis.com/maps/api/geocode/json?address=', 
                        address, '&sensor=false', sep = '')
    if(nchar(key) > 0){
      url_string <- paste(url_string, '&key=', key, sep = '')
    }
  }
  if(api == 'baidu'){
    # http://api.map.baidu.com/geocoder/v2/?address=ADDRESS&output=json&ak=API_KEY
    url_string <- paste('http://api.map.baidu.com/geocoder/v2/?address=', address, 
                        '&output=json&ak=', key, sep = '')
  }
  
  url_string <- URLencode(url_string)
  if(messaging) message(paste('calling ', url_string, ' ... ', sep = ''), appendLF = F)
  
  # gecode
  connect <- url(url_string)
  gc <- fromJSON(paste(readLines(connect, warn = FALSE), collapse = ''))
  if(messaging) message('done.')  
  close(connect)
  
  # geocoding results
  if(api == 'google'){
    # did geocode fail?
    if(gc$status != 'OK'){
      warning(paste('geocode failed with status ', gc$status, ', location = "', 
                    address, '"', sep = ''), call. = FALSE)
      return(data.frame(lat = NA, lng = NA))  
    }
    
    # more than one location found?
    if(length(gc$results) > 1 && messaging){
      Encoding(gc$results[[1]]$formatted_address) <- "UTF-8"
      message(paste('more than one location found for "', address, 
                    '", using address\n"', tolower(gc$results[[1]]$formatted_address), 
                    '"', sep = ''), appendLF = T)
    }
    
    gcdf <- with(gc$results[[1]], {
      data.frame(lat = NULLtoNA(geometry$location['lat']), 
                 lng = NULLtoNA(geometry$location['lng']), 
                 loctype = tolower(NULLtoNA(geometry$location_type)), 
                 row.names = NULL)})
    
    # convert coordinates only in China
    if(!outofChina(gcdf[, 'lat'], gcdf[, 'lng'])){
      gcdf[c('lat', 'lng')] <- conv(gcdf[, 'lat'], gcdf[, 'lng'], from = 'GCJ-02', to = ocs)
    } else{
      if(ocs != 'WGS-84'){
        message('wrong usage: for address out of China, ocs can only be set to "WGS-84"', appendLF = T)
      }
    }
    
    if(output == 'latlng') return(gcdf[c('lat', 'lng')])
    if(output == 'latlngc') return(gcdf[c('lat', 'lng', 'loctype')])
  }
  if(api == 'baidu'){
    # did geocode fail?
    if(gc$status != 0){
      warning(paste('geocode failed with status code ', gc$status, ', location = "', 
                    address, '". see more details in the response code table of Baidu Geocoding API', 
                    sep = ''), call. = FALSE)
      return(data.frame(lat = NA, lng = NA))
    }
    
    gcdf <- with(gc$result, {data.frame(lat = NULLtoNA(location['lat']), 
                                        lng = NULLtoNA(location['lng']), 
                                        conf = NULLtoNA(confidence), 
                                        row.names = NULL)})
    
    # convert coordinates
    gcdf[c('lat', 'lng')] <- conv(gcdf[, 'lat'], gcdf[, 'lng'], from = 'BD-09', to = ocs)
    
    if(output == 'latlng') return(gcdf[c('lat', 'lng')])
    if(output == 'latlngc') return(gcdf[c('lat', 'lng', 'conf')])
  }
}

# fill NULL with NA
NULLtoNA <- function(x){
  if(is.null(x)) return(NA)
  if(is.character(x) & nchar(x) == 0) return(NA)
  x
}

R/revgeocode.R
#' Reverse geocode
#'
#' reverse geocodes a lat/lng location using Google or Baidu Maps.  Note that in 
#' most cases by using this function you are agreeing to the Google Maps API Terms 
#' of Service at \url{https://developers.google.com/maps/terms} or the Baidu Maps 
#' API Terms of Use at \url{http://developer.baidu.com/map/law.htm}.
#' 
#' @param latlng a location in latitude/longitude format
#' @param ics the coordinate system of inputing location, including WGS-84, GCJ-02 
#' and BD-09, which are the GCSs of Google Earth, Google Map in China and Baidu 
#' Map, respectively. For location out of China, ics is automatically set to 'WGS-84'
#' and other values are ignored.
#' @param api use google or baidu maps api
#' @param key an api key must be provided when calling baidu maps api. 
#' While it's unnecessary for calling google maps api.
#' @param output formatted address or formmatted address with address components
#' @param messaging turn messaging on/off
#' @return a data.frame with variables address or detail address components 
#' @author Jun Cai (\email{cai-j12@@mails.tsinghua.edu.cn}), PhD student from 
#' Center for Earth System Science, Tsinghua University
#' @details note that the google maps api limits to 2500 queries a day.
#' @seealso \code{\link{geocode}}.
#' 
#' Google Maps API at \url{http://code.google.com/apis/maps/documentation/geocoding/} 
#' and Baidu Maps API at \url{http://developer.baidu.com/map/webservice-geocoding.htm}
#' @export
#' @examples
#' \dontrun{
#' # reverse geocode Beijing railway station
#' revgeocode(c(39.90105, 116.42079), ics = 'WGS-84', api = 'google', 
#'            output = 'address')
#' revgeocode(c(39.90245, 116.42703), ics = 'GCJ-02', api = 'google', 
#'            output = 'address', messaging = TRUE)
#' revgeocode(c(39.90851, 116.43351), ics = 'BD-09', api = 'google', 
#'            output = 'addressc')
#' revgeocode(c(39.90851, 116.43351), ics = 'BD-09', api = 'baidu', 
#'            key = 'your baidu maps api key', output = 'address')
#' revgeocode(c(39.90245, 116.42703), ics = 'GCJ-02', api = 'baidu', 
#'            key = 'your baidu maps api key', output = 'address', messaging = TRUE)
#' revgeocode(c(39.90105, 116.42079), ics = 'WGS-84', api = 'baidu', 
#'            key = 'your baidu maps api key', output = 'addressc')
#' # reverse geocode multiple locations
#' latlng = data.frame(lat = c(39.99837, 39.98565), lng = c(116.3203, 116.2998))
#' revgeocode(latlng, ics = 'WGS-84', api = 'google', output = 'address')
#' }

revgeocode <- function(latlng, ics = c('WGS-84', 'GCJ-02', 'BD-09'), 
                       api = c('google', 'baidu'), key = '', 
                       output = c('address', 'addressc'), messaging = FALSE){
  # check parameters
  stopifnot(class(latlng) %in% c('numeric', 'data.frame'))
  ics <- match.arg(ics)
  api <- match.arg(api)
  stopifnot(is.character(key))
  output <- match.arg(output)
  stopifnot(is.logical(messaging))
  
  if(is.data.frame(latlng)){
    return(ldply(seq_along(latlng), function(i){ revgeocode(as.numeric(latlng[i, ]), 
                                                            ics = ics, api = api, 
                                                            key = key, output = output, 
                                                            messaging = messaging) }))
  }
  
  # format url
  if(api == 'google'){
    # convert coordinates only in China
    if(!outofChina(latlng[1], latlng[2])){
      latlng <- conv(latlng[1], latlng[2], from = ics, to = 'GCJ-02')
    } else{
      if(ics != 'WGS-84'){
        message('wrong usage: for location out of China, ics can only be set to "WGS-84"', appendLF = T)
      }
    }
    
    # http://maps.googleapis.com/maps/api/geocode/json?latlng=LAT,LNG
    # &sensor=FALSE&key=API_KEY
    url_string <- paste('http://maps.googleapis.com/maps/api/geocode/json?latlng=', 
                        latlng[1], ',', latlng[2], '&sensor=false', sep = '')
    if(nchar(key) > 0){
      url_string <- paste(url_string, '&key=', key, sep = '')
    }
  }
  if(api == 'baidu'){
    # coordinate type lookup table
    code <- c('wgs84ll', 'gcj02ll', 'bd09ll')
    names(code) <- c('WGS-84', 'GCJ-02', 'BD-09')
    coordtype <- code[ics]
    # http://api.map.baidu.com/geocoder/v2/?location=LAT,LNG&coordtype=COORDTYPE
    # &output=json&ak=API_KEY
    url_string <- paste('http://api.map.baidu.com/geocoder/v2/?location=', 
                        latlng[1], ',', latlng[2], '&coordtype=', coordtype, 
                        '&output=json&ak=', key, sep = '')
  }
  
  url_string <- URLencode(url_string)
  if(messaging) message(paste('calling ', url_string, ' ... ', sep = ''), appendLF = F)
  
  # reverse gecode
  connect <- url(url_string)
  rgc <- fromJSON(paste(readLines(connect, warn = FALSE), collapse = ''))
  if(messaging) message('done.')  
  close(connect)
  
  # reverse geocoding results
  if(api == 'google'){
    # did reverse geocoding fail?
    if(rgc$status != 'OK'){
      warning(paste('reverse geocode failed with status ', gc$status, ', location = "', 
                    latlng[1], ', ', latlng[2], '"', sep = ''), call. = FALSE)
      return(data.frame(address = NA))  
    }
    
    # more than one address found?
    if(length(rgc$results) > 1 && messaging){
      message(paste('more than one address found for "', latlng[1], ', ', 
                    latlng[2],  '", reverse geocoding first ...', sep = ''), apppendLF = T)
    }
    
    rgcdf <- with(rgc$results[[1]], {data.frame(address = formatted_address, 
                                                row.names = NULL)})
    for(i in seq_along(rgc$results[[1]]$address_components)){
      rgcdf <- cbind(rgcdf, rgc$results[[1]]$address_components[[i]]$long_name)
    }
    names(rgcdf) <- c('address', sapply(rgc$results[[1]]$address_components, 
                                        function(l) l$types[1]))
  }
  if(api == 'baidu'){
    # did geocode fail?
    if(rgc$status != 0){
      warning(paste('geocode failed with status code ', rgc$status, ', location = "', 
                    latlng[1], ', ', latlng[2],  '". see more details in the response code table of Baidu Geocoding API', 
                    sep = ''), call. = FALSE)
      return(data.frame(address = NA))
    }
    
    rgcdf <- with(rgc$result, {
      data.frame(address = formatted_address, 
                 street_number = NULLtoNA(addressComponent['street_number']), 
                 street = NULLtoNA(addressComponent['street']), 
                 district = NULLtoNA(addressComponent['district']), 
                 city = NULLtoNA(addressComponent['city']), 
                 province = NULLtoNA(addressComponent['province']), 
                 row.names = NULL)})
  }
  
  if(output == 'address') return(rgcdf['address'])
  if(output == 'addressc') return(rgcdf)
}